core/src/main/java/org/apache/calcite/rel/rules/FilterJoinRule.java

```diff
@@ -329,7 +330,14 @@ protected void validateJoinFilters(List<RexNode> aboveFilters,
     final Iterator<RexNode> filterIter = joinFilters.iterator();
     while (filterIter.hasNext()) {
       RexNode exp = filterIter.next();
-      if (!predicate.apply(join, joinType, exp)) {
+      int maxIndex = -1;
+      ImmutableList<RexInputRef> inputRefs = RexUtil.getAllInputRef(exp);
```

Comment:
Rather than extracting out all the inputRefs and compare field count of join with the biggest index of inputRefs, shall we just check all the inputRefs by RexVisitor and jump out (by throw-catch) when finding an inputRef which has index bigger than field count of join.

Reply:
Or shall we just check the join type by `JoinRelType#projectsRight` ?

Reply:
the join type is already check in `predicate.apply(join, joinType, exp)`, so I would prefer not to check again.

Reply:
No, it doesn't.